File: Graph/TopologicalSort/TopologicalSort.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def topological_sort_util(self, vertex, visited, stack):
        visited.add(vertex)

        for neighbor in self.graph[vertex]:
            if neighbor not in visited:
                self.topological_sort_util(neighbor, visited, stack)

        stack.append(vertex)

    def topological_sort(self):
        visited = set()
        stack = []

        for vertex in self.graph:
            if vertex not in visited:
                self.topological_sort_util(vertex, visited, stack)

        return stack[::-1]


# Пример использования
graph = Graph()
graph.add_edge(1, 2)
graph.add_edge(1, 3)
graph.add_edge(2, 4)
graph.add_edge(2, 5)
graph.add_edge(3, 6)
graph.add_edge(3, 7)

result = graph.topological_sort()
print("Topological Sort:", result)

File: Graph/TopologicalSort/TopologicalSort.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Graph
{
    private Dictionary<int, List<int>> graph = new Dictionary<int, List<int>>();

    public void AddEdge(int u, int v)
    {
        if (!graph.ContainsKey(u))
        {
            graph[u] = new List<int>();
        }

        graph[u].Add(v);
    }

    private void TopologicalSortUtil(int vertex, HashSet<int> visited, Stack<int> stack)
    {
        visited.Add(vertex);

        foreach (int neighbor in graph[vertex])
        {
            if (!visited.Contains(neighbor))
            {
                TopologicalSortUtil(neighbor, visited, stack);
            }
        }

        stack.Push(vertex);
    }

    public List<int> TopologicalSort()
    {
        HashSet<int> visited = new HashSet<int>();
        Stack<int> stack = new Stack<int>();
        List<int> result = new List<int>();

        foreach (var entry in graph)
        {
            int vertex = entry.Key;
            if (!visited.Contains(vertex))
            {
                TopologicalSortUtil(vertex, visited, stack);
            }
        }

        while (stack.Count > 0)
        {
            result.Add(stack.Pop());
        }

        return result;
    }
}

class Program
{
    static void Main()
    {
        Graph graph = new Graph();
        graph.AddEdge(1, 2);
        graph.AddEdge(1, 3);
        graph.AddEdge(2, 4);
        graph.AddEdge(2, 5);
        graph.AddEdge(3, 6);
        graph.AddEdge(3, 7);

        List<int> result = graph.TopologicalSort();

        Console.Write("Topological Sort: ");
        foreach (int vertex in result)
        {
            Console.Write($"{vertex} ");
        }
    }
}